### ISSUE TITLE
Effort to support integral underlying types

### ISSUE THREAD

**Issue by johelegp** [nholthaus/units#125]
There is interest in having the library support integral underlying types. I believe that the [`duration`](https://wg21.link/time.duration) class template of [`std::chrono`](https://wg21.link/time.syn) sets a good example in doing so. Its interface is correct, efficient, easy to use right, and hard to misuse. I propose that we adopt the ways of `std::chrono::duration` to make this library more generally useful.

To properly support integral underlying types, the following points need to be implemented. First comes the starting point of the `unit` objects lifetime, the constructors and its dependent utilities. Then `std::common_type`, the base of safe conversions between units of the same dimensions.

- The underlying type parameter of `unit` constructors [[1](https://github.com/nholthaus/units/blob/9b66a12e42e9dffde3be8c0d3eab2bbe5bd0298a/include/units.h#L1908), [2](https://github.com/nholthaus/units/blob/9b66a12e42e9dffde3be8c0d3eab2bbe5bd0298a/include/units.h#L1919)] should be a template parameter of the constructors that only allows [non-lossy conversions](https://wg21.link/time.duration.cons#1) [ _Note:_ `duration` uses `treat_­as_­floating_­point_­v` because it supports arithmetic-like types, for which the standard traits can't be specialized. To keep this effort simple, I think we should continue using the standard traits. -- _end note_ ].
- The [converting constructor](https://github.com/nholthaus/units/blob/9b66a12e42e9dffde3be8c0d3eab2bbe5bd0298a/include/units.h#L1942) should [prevent truncation and lossy conversions](https://wg21.link/time.duration.cons#4).
- The [`convert`](https://github.com/nholthaus/units/blob/9b66a12e42e9dffde3be8c0d3eab2bbe5bd0298a/include/units.h#L1561) function, used by the converting constructor:
  * Needs to [carry intermediate computations in the widest representation](https://wg21.link/time.duration.cast#3).
  * Should encourage type safety by having a type safe interface, or an alternative provided that has a type safe interface. A type safe interface would have `unit`s in the parameter and return types. If an alternative is provided, `convert` should work with `unit`s instead of `unit_conversion`s because `unit`s know the specific underlying types being handled, allowing `convert` to have a correctly specified parameter and return types.
- Partially specialize [`std::common_type`](https://wg21.link/time.traits.specializations). This is pretty straight forward for linear units that don't need a translation, just like `duration`. I'm not sure how to go about it for other units.
- Provide support for the modulus operators.

This list is a work in progress. I'm just letting this out so that discussion can start.

**Comment by nholthaus** [comment]
I'll add that if we get this right, we should also include definitions for additional underlying type units.

**Comment by johelegp** [comment]
>we should also include definitions for additional underlying type units.

What do you mean by that? Like how the standard `0s` is an integral duration and `.0s` a floating-point duration?

**Comment by johelegp** [comment]
I think that's about it. The main comment should be complete now.

**Comment by nholthaus** [comment]
> What do you mean by that? Like how the standard 0s is an integral duration and .0s a floating-point duration?

I just mean defining types/alias like `meter_t`, only with an `int` underlying type. I'm planning to remove the `_t`'s as part of the 3.0 refactor (although I'd gladly accept user feedback on that), so I'm not exactly sure what I'd want to call them. I think defining a single underlying type for the entire lib is too limiting, especially given your recent work.

I don't think the duration model will quite work for us because I still think the default/most readily accessible underlying type should be floating point.

**Comment by johelegp** [comment]
>I just mean defining types/alias like meter_t, only with an int underlying type.

That'd be nice to have.
>I'm planning to remove the _t's as part of the 3.0 refactor

That's great. I've read some things about that in previous issues and was planning to mention it to you because I didn't see anything about it for 3.0.0.
>(although I'd gladly accept user feedback on that)

Then you'll want to open an issue for visibility.

>I don't think the duration model will quite work for us because I still think the default/most readily accessible underlying type should be floating point.

The effort changes nothing in that respect.

**Comment by nholthaus** [comment]
Are we ready to close this?

**Comment by johelegp** [comment]
Yeah. It seems like you've disabled [closing issues from PRs](https://help.github.com/articles/closing-issues-using-keywords/).

**Comment by nholthaus** [comment]
Right. I don't want them fully closed until they merge into master, but I
like to tag them so I know that no more active quirk is required.

**Comment by nholthaus** [comment]
Fixed in 3.x: integral underlying types are supported and tested (`meters<int>`), documented in the README. Closing.
